File: app/routes/links.py
"""Player-profile linking: user requests + admin approval."""
from __future__ import annotations

from datetime import datetime

from flask import Blueprint, jsonify, request
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt

from app import db, limiter
from app.models import LinkRequest, Player, User
from app.routes.user_auth import current_user, user_required

links_bp = Blueprint("links", __name__)


def _admin_required(fn):
    """Admin-only route guard (claim role != 'user')."""
    from functools import wraps

    @wraps(fn)
    @jwt_required()
    def wrapper(*args, **kwargs):
        if get_jwt().get("role") == "user":
            return jsonify({"error": "Admin access required"}), 403
        return fn(*args, **kwargs)

    return wrapper


# ---- USER endpoints ---------------------------------------------------------

@links_bp.route("/request", methods=["POST"])
@user_required
@limiter.limit("5 per hour")
def request_link():
    u = current_user()
    if u.linked_player_id:
        return jsonify({"error": "Account is already linked to a player"}), 400

    data = request.get_json() or {}
    player_id = data.get("player_id")
    message = (data.get("message") or "").strip()[:1000] or None
    if not player_id:
        return jsonify({"error": "player_id is required"}), 400

    player = Player.query.get(int(player_id))
    if not player:
        return jsonify({"error": "Player not found"}), 404

    # Reject if a pending request already exists for this user
    existing = LinkRequest.query.filter_by(user_id=u.id, status="pending").first()
    if existing:
        return jsonify({"error": "You already have a pending link request",
                        "request": existing.to_dict()}), 409

    # Reject if this player is already linked to another user
    other = User.query.filter_by(linked_player_id=player.id).first()
    if other:
        return jsonify({"error": "This player is already linked to another account"}), 409

    lr = LinkRequest(user_id=u.id, player_id=player.id, message=message, status="pending")
    db.session.add(lr)
    db.session.commit()
    return jsonify(lr.to_dict()), 201


@links_bp.route("/my-requests", methods=["GET"])
@user_required
def my_requests():
    lang = request.args.get("lang", "en")
    rows = LinkRequest.query.filter_by(user_id=current_user().id).order_by(
        LinkRequest.created_at.desc()).all()
    return jsonify([r.to_dict(lang) for r in rows]), 200


# ---- ADMIN endpoints --------------------------------------------------------

@links_bp.route("/admin/requests", methods=["GET"])
@_admin_required
def list_requests():
    lang = request.args.get("lang", "en")
    status = request.args.get("status")  # pending | approved | rejected | all
    q = LinkRequest.query
    if status and status != "all":
        q = q.filter(LinkRequest.status == status)
    rows = q.order_by(LinkRequest.created_at.desc()).limit(200).all()
    return jsonify([r.to_dict(lang) for r in rows]), 200


@links_bp.route("/admin/requests/<int:req_id>/approve", methods=["POST"])
@_admin_required
def approve_request(req_id: int):
    data = request.get_json() or {}
    lr = LinkRequest.query.get_or_404(req_id)
    if lr.status != "pending":
        return jsonify({"error": "Already reviewed"}), 400

    # Re-check the player isn't already linked
    other = User.query.filter(User.linked_player_id == lr.player_id,
                              User.id != lr.user_id).first()
    if other:
        return jsonify({"error": "Player is already linked to another account"}), 409

    lr.status = "approved"
    lr.admin_note = (data.get("admin_note") or "").strip()[:1000] or None
    lr.reviewed_by_admin_id = int(get_jwt_identity())
    lr.reviewed_at = datetime.utcnow()
    lr.user.linked_player_id = lr.player_id
    db.session.commit()
    return jsonify(lr.to_dict()), 200


@links_bp.route("/admin/requests/<int:req_id>/reject", methods=["POST"])
@_admin_required
def reject_request(req_id: int):
    data = request.get_json() or {}
    lr = LinkRequest.query.get_or_404(req_id)
    if lr.status != "pending":
        return jsonify({"error": "Already reviewed"}), 400
    lr.status = "rejected"
    lr.admin_note = (data.get("admin_note") or "").strip()[:1000] or None
    lr.reviewed_by_admin_id = int(get_jwt_identity())
    lr.reviewed_at = datetime.utcnow()
    db.session.commit()
    return jsonify(lr.to_dict()), 200


@links_bp.route("/admin/users/<int:user_id>/unlink", methods=["POST"])
@_admin_required
def admin_unlink(user_id: int):
    """Admin can break a link if it was wrongly granted."""
    u = User.query.get_or_404(user_id)
    u.linked_player_id = None
    db.session.commit()
    return jsonify(u.public_dict()), 200


# ---- ADMIN: user moderation ------------------------------------------------

@links_bp.route("/admin/users", methods=["GET"])
@_admin_required
def list_users():
    """List all platform users with filters for moderation."""
    status = request.args.get("status", "all")  # all | active | banned | unverified
    search = (request.args.get("search") or "").strip()
    page = max(1, int(request.args.get("page", 1)))
    per_page = min(100, max(1, int(request.args.get("per_page", 25))))

    q = User.query
    if status == "banned":
        q = q.filter(User.is_banned.is_(True))
    elif status == "active":
        q = q.filter(User.is_banned.is_(False), User.is_verified.is_(True))
    elif status == "unverified":
        q = q.filter(User.is_verified.is_(False))

    if search:
        like = f"%{search}%"
        q = q.filter(
            (User.username.ilike(like))
            | (User.email.ilike(like))
            | (User.display_name.ilike(like))
        )

    q = q.order_by(User.created_at.desc())
    paged = q.paginate(page=page, per_page=per_page, error_out=False)
    return jsonify({
        "users": [u.private_dict() for u in paged.items],
        "total": paged.total,
        "page": page,
        "pages": paged.pages,
        "per_page": per_page,
    }), 200


@links_bp.route("/admin/users/<int:user_id>/ban", methods=["POST"])
@_admin_required
def ban_user(user_id: int):
    data = request.get_json() or {}
    u = User.query.get_or_404(user_id)
    u.is_banned = True
    u.banned_at = datetime.utcnow()
    u.ban_reason = (data.get("reason") or "").strip()[:1000] or None
    db.session.commit()
    return jsonify(u.private_dict()), 200


@links_bp.route("/admin/users/<int:user_id>/unban", methods=["POST"])
@_admin_required
def unban_user(user_id: int):
    u = User.query.get_or_404(user_id)
    u.is_banned = False
    u.banned_at = None
    u.ban_reason = None
    db.session.commit()
    return jsonify(u.private_dict()), 200


@links_bp.route("/admin/users/<int:user_id>/verify", methods=["POST"])
@_admin_required
def admin_verify_user(user_id: int):
    """Admin can manually mark a user verified (e.g. failed email delivery)."""
    u = User.query.get_or_404(user_id)
    u.is_verified = True
    u.otp_code = None
    u.otp_expires_at = None
    db.session.commit()
    return jsonify(u.private_dict()), 200


@links_bp.route("/admin/users/<int:user_id>/mute", methods=["POST"])
@_admin_required
def admin_mute_user(user_id: int):
    """Admin chat-mute (prevents user from posting any chat / DM)."""
    u = User.query.get_or_404(user_id)
    u.chat_muted = True
    db.session.commit()
    return jsonify(u.private_dict()), 200


@links_bp.route("/admin/users/<int:user_id>/unmute", methods=["POST"])
@_admin_required
def admin_unmute_user(user_id: int):
    u = User.query.get_or_404(user_id)
    u.chat_muted = False
    db.session.commit()
    return jsonify(u.private_dict()), 200

File: server/src/routes/links.ts
import { Router } from 'express';
import { z } from 'zod';
import { LinkRequest, Player, User } from '../models/index.js';
import { asyncHandler } from '../lib/async-handler.js';
import { HttpError } from '../lib/http-error.js';
import { langQuery, objectId, pagination, parseBody, parseQuery } from '../lib/validate.js';
import {
  normalizeLang,
  paginationMeta,
  serializePlayer,
  serializeUser,
  serializeUserPrivate,
  type Lang,
} from '../lib/serializers.js';
import { escapeRegex, trimToNull } from '../lib/sanitize.js';
import { currentAdmin, currentUser, requireAdmin, requireUser } from '../middleware/auth.js';
import { linkRequestLimiter } from '../middleware/rate-limit.js';
import { notifyUser } from '../realtime/publish.js';

export const linksRouter: Router = Router();

function serializeLinkRequest(
  request: {
    _id: unknown;
    userId: unknown;
    playerId: unknown;
    message?: string | null;
    status: string;
    adminNote?: string | null;
    reviewedAt?: Date | null;
    createdAt?: Date;
  },
  lang: Lang,
) {
  const user = request.userId as unknown;
  const player = request.playerId as unknown;
  const isDoc = (value: unknown): value is Record<string, unknown> =>
    typeof value === 'object' && value !== null && '_id' in value && Object.keys(value).length > 1;

  return {
    id: String(request._id),
    user_id: isDoc(user) ? String(user._id) : String(user),
    user: isDoc(user) ? serializeUser(user as never) : null,
    player_id: isDoc(player) ? String(player._id) : String(player),
    player: isDoc(player) ? serializePlayer(player as never, lang) : null,
    message: request.message ?? null,
    status: request.status,
    admin_note: request.adminNote ?? null,
    reviewed_at: request.reviewedAt ? new Date(request.reviewedAt).toISOString() : null,
    created_at: request.createdAt ? new Date(request.createdAt).toISOString() : null,
  };
}

// -------------------------------------------------------------------- user

const requestSchema = z.object({
  player_id: objectId,
  message: z.string().max(1000).optional(),
});

linksRouter.post(
  '/request',
  requireUser,
  linkRequestLimiter,
  asyncHandler(async (req, res) => {
    const { player_id: playerId, message } = parseBody(requestSchema, req);
    const user = currentUser(req);

    if (user.linkedPlayerId) throw HttpError.badRequest('Your account is already linked to a profile');

    const player = await Player.findById(playerId);
    if (!player) throw HttpError.notFound('Player not found');

    if (await User.exists({ linkedPlayerId: playerId })) {
      throw HttpError.conflict('That profile is already linked to another account');
    }

    try {
      const request = await LinkRequest.create({
        userId: user._id,
        playerId,
        message: trimToNull(message, 1000),
        status: 'pending',
      });
      await request.populate('playerId');
      res.status(201).json(serializeLinkRequest(request, normalizeLang(req.query.lang)));
    } catch (error) {
      // The unique partial index is what actually enforces one open request
      // per user, so two simultaneous submissions cannot both get through.
      if ((error as { code?: number }).code === 11000) {
        throw HttpError.conflict('You already have a pending link request');
      }
      throw error;
    }
  }),
);

linksRouter.get(
  '/my-requests',
  requireUser,
  asyncHandler(async (req, res) => {
    const { lang } = parseQuery(langQuery, req);
    const requests = await LinkRequest.find({ userId: currentUser(req)._id })
      .sort({ createdAt: -1 })
      .limit(50)
      .populate('playerId');

    res.json(requests.map((request) => serializeLinkRequest(request, lang)));
  }),
);

// --------------------------------------------------------- admin: requests

const adminListQuery = z.object({
  status: z.enum(['pending', 'approved', 'rejected', 'all']).default('all'),
  lang: z.enum(['en', 'ar']).default('en'),
});

linksRouter.get(
  '/admin/requests',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { status, lang } = parseQuery(adminListQuery, req);

    const requests = await LinkRequest.find(status === 'all' ? {} : { status })
      .sort({ createdAt: -1 })
      .limit(200)
      .populate('userId')
      .populate('playerId');

    res.json(requests.map((request) => serializeLinkRequest(request, lang)));
  }),
);

const reviewSchema = z.object({ admin_note: z.string().max(1000).optional() });

linksRouter.post(
  '/admin/requests/:id/approve',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const { admin_note: note } = parseBody(reviewSchema, req);
    const admin = currentAdmin(req);

    const request = await LinkRequest.findById(id);
    if (!request) throw HttpError.notFound('Request not found');
    if (request.status !== 'pending') throw HttpError.badRequest('That request has already been reviewed');

    // The link is what grants identity, so set it first and only record the
    // approval if the write succeeds. The unique index on linkedPlayerId means
    // a profile can never end up backing two accounts.
    const linked = await User.updateOne(
      { _id: request.userId, linkedPlayerId: null },
      { $set: { linkedPlayerId: request.playerId } },
    );

    if (linked.modifiedCount === 0) {
      const already = await User.findById(request.userId).select('linkedPlayerId');
      if (already?.linkedPlayerId) {
        throw HttpError.conflict('That account is already linked to a profile');
      }
      throw HttpError.conflict('Could not link that account');
    }

    request.status = 'approved';
    request.adminNote = trimToNull(note, 1000);
    request.reviewedByAdminId = admin._id;
    request.reviewedAt = new Date();
    await request.save();

    await request.populate('userId');
    await request.populate('playerId');

    notifyUser(String(request.userId), 'link:reviewed', { status: 'approved' });
    res.json(serializeLinkRequest(request, normalizeLang(req.query.lang)));
  }),
);

linksRouter.post(
  '/admin/requests/:id/reject',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const { admin_note: note } = parseBody(reviewSchema, req);

    const request = await LinkRequest.findById(id);
    if (!request) throw HttpError.notFound('Request not found');
    if (request.status !== 'pending') throw HttpError.badRequest('That request has already been reviewed');

    request.status = 'rejected';
    request.adminNote = trimToNull(note, 1000);
    request.reviewedByAdminId = currentAdmin(req)._id;
    request.reviewedAt = new Date();
    await request.save();

    await request.populate('userId');
    await request.populate('playerId');

    notifyUser(String(request.userId), 'link:reviewed', { status: 'rejected' });
    res.json(serializeLinkRequest(request, normalizeLang(req.query.lang)));
  }),
);

// ------------------------------------------------------------ admin: users

const usersQuery = pagination(25).extend({
  status: z.enum(['all', 'active', 'banned', 'unverified']).default('all'),
  search: z.string().max(100).optional(),
});

linksRouter.get(
  '/admin/users',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { page, per_page: perPage, status, search } = parseQuery(usersQuery, req);

    const filter: Record<string, unknown> = {};
    if (status === 'banned') filter.isBanned = true;
    if (status === 'active') Object.assign(filter, { isBanned: false, isVerified: true });
    if (status === 'unverified') filter.isVerified = false;

    if (search?.trim()) {
      const pattern = new RegExp(escapeRegex(search.trim()), 'i');
      filter.$or = [{ username: pattern }, { email: pattern }, { displayName: pattern }];
    }

    const [users, total] = await Promise.all([
      User.find(filter)
        .sort({ createdAt: -1 })
        .skip((page - 1) * perPage)
        .limit(perPage)
        .populate('linkedPlayerId', 'nameEn title'),
      User.countDocuments(filter),
    ]);

    res.json({
      users: users.map(serializeUserPrivate),
      ...paginationMeta({ page, perPage, total }),
    });
  }),
);

/** Apply a moderation change and return the updated account. */
async function moderate(userId: string, update: Record<string, unknown>) {
  const user = await User.findByIdAndUpdate(userId, update, { returnDocument: 'after' }).populate(
    'linkedPlayerId',
    'nameEn title',
  );
  if (!user) throw HttpError.notFound('Player not found');
  return user;
}

const banSchema = z.object({ reason: z.string().max(1000).optional() });

linksRouter.post(
  '/admin/users/:id/ban',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const { reason } = parseBody(banSchema, req);

    const user = await moderate(id, {
      $set: { isBanned: true, bannedAt: new Date(), banReason: trimToNull(reason, 1000) },
    });

    notifyUser(id, 'account:banned', { reason: user.banReason });
    res.json(serializeUserPrivate(user));
  }),
);

linksRouter.post(
  '/admin/users/:id/unban',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const user = await moderate(id, {
      $set: { isBanned: false, bannedAt: null, banReason: null },
    });
    res.json(serializeUserPrivate(user));
  }),
);

linksRouter.post(
  '/admin/users/:id/verify',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    const user = await moderate(id, {
      $set: { isVerified: true, otpCodeHash: null, otpExpiresAt: null, otpAttempts: 0 },
    });
    res.json(serializeUserPrivate(user));
  }),
);

linksRouter.post(
  '/admin/users/:id/mute',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    res.json(serializeUserPrivate(await moderate(id, { $set: { chatMuted: true } })));
  }),
);

linksRouter.post(
  '/admin/users/:id/unmute',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    res.json(serializeUserPrivate(await moderate(id, { $set: { chatMuted: false } })));
  }),
);

linksRouter.post(
  '/admin/users/:id/unlink',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = objectId.parse(req.params.id);
    res.json(serializeUserPrivate(await moderate(id, { $set: { linkedPlayerId: null } })));
  }),
);
